[SecuritySolution] Dashboard listing UI

angorayc opened the pull request:
## Summary


1. Align dashboard listing UI with Kibana dashboard.
2. `Security Solution` tags are selected by default and removable by users.

**Prerequisite:**
This PR is waiting for https://github.com/elastic/kibana/pull/160871 to be merged


**Steps to verify:**
1. Visit Security > Dashboards, and create a dashboard from this page.
2. Back to Security Dashboards page, you should see the dashboard you just created and Security Solution tag should be selected by default in the tag filters.
3. Open the tag options, click the Security Solution tag. Observe that it should be removable, and it should display all the dashboards you have in the table.

**Known issues:**
https://github.com/elastic/kibana/pull/160540#issuecomment-1610395834

**Before:**

<img width="2545" alt="Screenshot 2023-06-27 at 09 24 19" src="https://github.com/elastic/kibana/assets/6295984/bc0fa0b1-96ad-43b0-afc1-48444dfb5691">


**After:**
<img width="2543" alt="Screenshot 2023-06-27 at 09 22 21" src="https://github.com/elastic/kibana/assets/6295984/82d0a868-bda2-431f-b0b5-9cbc34d3ae71">


### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

angorayc committed:
styling

angorayc committed:
add fixed tag

angorayc committed:
disable actions

angorayc committed:
replace dashboard listing

angorayc committed:
default selected tags

angorayc commented on packages/content-management/table_list_view_table/src/components/use_tag_filter_panel.tsx:
```diff
@@ -52,8 +55,26 @@ export const useTagFilterPanel = ({
+        }, {} as TagSelection)
+      : {}
+  );
+  const totalActiveFilters = useMemo(
+    () =>
+      Object.keys(tagSelection).reduce((acc, currentOption) => {
+        const inTagReferences = tagReferences?.find((ref) => ref.name === currentOption);
+        acc +=
+          inTagReferences != null
+            ? tagReferences?.filter((ref) => ref.name === currentOption).length ?? 0
+            : 1;
+        return acc;
+      }, 0),
+    [tagReferences, tagSelection]
+  );
```
This is to fix this scenario:
<img width="515" alt="incorrect_tag_counts" src="https://github.com/elastic/kibana/assets/6295984/2143b4b2-bb9a-41f7-8fd7-e91811504e75">

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

angorayc committed:
fix tag types

angorayc commented on packages/content-management/table_list_view_table/src/components/table.tsx:
```diff
@@ -150,7 +152,17 @@ export function Table<T extends UserContentCommonSchema>({
   } = useTagFilterPanel({
     query: searchQuery.query,
-    getTagList,
+    getTagList: useCallback(() => {
+      const tags = getTagList();
+      // In some situation, tags are created after page load.
+      // Therefore the getTagList result is cached before tags created
+      // Do this check to make sure default selected tags are showed
+      if (tags.length === 0 && tagReferences != null) {
+        return tagReferences;
+      }
+      return tags;
+    }, [getTagList, tagReferences]),
```
If we don't do the check, it returns the value cached on page load, but there are some scenario that tags create after page loaded:
<img width="2555" alt="incorrect_cache_tag_count" src="https://github.com/elastic/kibana/assets/6295984/7aaab121-7acc-40e8-b8a0-78efd4f07c98">

After adding the check for initial tag references:
<img width="2545" alt="corrected_tag_count" src="https://github.com/elastic/kibana/assets/6295984/4ff9e3de-470d-4c2c-9c44-72438c05b1e6">

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

angorayc committed:
fix types

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

elasticmachine commented:
Pinging @elastic/security-threat-hunting (Team:Threat Hunting)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

angorayc committed:
unit tests

sebelga reviewed (changes requested):
Thanks for working on this @angorayc 

I am sorry but I need to push back changes related to "layout". You've added 3 new props for it:

* `withPageTemplateHeader`
* `restrictPageSectionWidth`
* `pageSectionPadding`

which are all related to changes in layout.

Since https://github.com/elastic/kibana/pull/157988 there is a `<TableListViewTable />` that is exposed and only contains the `Table` which seems to be what you need.

It would probably means that the `DashboardListing` should be refactored to only export the Table for Dashboard as well.

If you want we can discuss over Zoom 👍

angorayc committed:
consume TableListViewTable from dashboard listing

sebelga commented on packages/content-management/table_list_view_table/src/services.tsx:
```diff
@@ -201,7 +201,10 @@ export const TableListViewKibanaProvider: FC<TableListViewKibanaDependencies> =
 
   const getTagList = useCallback(() => {
     if (!savedObjectsTagging?.ui.getTagList) {
```
This should be `savedObjectsTagging?.ui.getTagList$`

sebelga commented on packages/content-management/table_list_view_table/src/services.tsx:
```diff
@@ -201,7 +201,10 @@ export const TableListViewKibanaProvider: FC<TableListViewKibanaDependencies> =
   const getTagList = useCallback(() => {
     if (!savedObjectsTagging?.ui.getTagList) {
-      return [];
+      return new Observable<Tag[]>((subscriber) => {
```
Here you want to use `of`

```ts
return of([]);
```

sebelga commented on packages/content-management/table_list_view_table/src/table_list_view_table.tsx:
```diff
@@ -861,41 +861,43 @@ function TableListViewTableComp<T extends UserContentCommonSchema>({
         termMatch = searchTerm;
 
         if (references?.length || referencesToExclude?.length) {
-          const allTags = getTagList();
```
This should be replace at the top of the component with

```ts
const allTags = useObservable(getTagList());
```

sebelga commented on src/plugins/saved_objects_tagging_oss/public/api.ts:
```diff
@@ -71,6 +71,8 @@ export interface SavedObjectsTaggingApiUi {
    */
   getTagList(): Tag[];
 
+  getTagList$: Observable<Tag[]>;
```
If it's a function it should be

```ts
getTagList$(): Observable<Tag[]>;
```

sebelga reviewed (commented):
Left some comments

angorayc committed:
revert tagReferences

angorayc committed:
fix filter count

angorayc committed:
rename

angorayc commented on packages/content-management/table_list_view_table/src/reducer.tsx:
```diff
@@ -63,7 +64,10 @@ export function getReducer<T extends UserContentCommonSchema>() {
       }
       case 'onSearchQueryChange': {
-        if (action.data.text === state.searchQuery.text) {
+        const clause = state.searchQuery.query.ast.clauses;
+        const newClause = action.data.query.ast.clauses;
```
Without checking the change between clauses leads to querySearch not updated:


https://github.com/elastic/kibana/assets/6295984/0a81ab10-d7e5-4629-bfff-d7bde4efe7c2

angorayc replied:
On the page load, we can see `tag("Security Solution")` in the search box, but non of the tag filter is selected. Expect that both Security Solution should be selected and shows the correct active count.

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

angorayc commented on packages/content-management/table_list_view_table/src/table_list_view_table.tsx:
```diff
@@ -877,9 +878,15 @@ function TableListViewTableComp<T extends UserContentCommonSchema>({
+      updateSortFromURL(urlState.sort);
+    }
+
+    if (initialQuery) {
```
This is to align initialFilter and initial selected tags as well. Otherwise the searchQuery is incorrect as we use Ast: [] to init for the initial query.

angorayc commented on packages/content-management/table_list_view_table/src/components/use_tag_filter_panel.tsx:
```diff
@@ -45,15 +45,43 @@ export const useTagFilterPanel = ({
+    }, {} as TagSelection),
+  });
+
+  const totalActiveFilters = useMemo(
+    () =>
+      tags.filter((currentOption) =>
+        [...clauseInclude, ...clauseExclude].find((clauseName) => clauseName === currentOption.name)
+      ).length,
+    [clauseExclude, clauseInclude, tags]
+  );
```
Before:
<img width="2540" alt="incorrect_active_count" src="https://github.com/elastic/kibana/assets/6295984/2fe07160-1f9e-498e-a683-5d64729fc88b">


After:
https://github.com/elastic/kibana/assets/6295984/63fd6d2e-2a07-45c8-a624-a50ca8d8b3bc

angorayc replied:
Reverted the change in this PR. Follow up in https://github.com/elastic/kibana/issues/160720

angorayc commented on src/plugins/dashboard/public/dashboard_listing/dashboard_listing.tsx:
```diff
@@ -225,35 +254,56 @@ export const DashboardListing = ({
+                  )
+                }
+              />
+              <TableListViewTable
```
Shared UX team prefer not to add more props to `TableListView`, so suggested us to use `TableListViewTable` to avoid adding more new props.

angorayc committed:
styling

angorayc committed:
Merge branch 'dashboard-listing' of github.com:angorayc/kibana into dashboard-listing

angorayc commented:
Known issue: getTagList returns incorrect result if tags are create after page loaded:


<img width="2550" alt="incorrect_getList" src="https://github.com/elastic/kibana/assets/6295984/c80efc7a-e442-456c-a327-c3a832753ffd">




https://github.com/elastic/kibana/assets/6295984/74f16302-7abc-4fc1-869a-4a7b0bf54556

angorayc committed:
unit test

sebelga commented on packages/content-management/table_list_view_table/src/components/use_tag_filter_panel.tsx:
```diff
@@ -45,15 +45,43 @@ export const useTagFilterPanel = ({
   const [options, setOptions] = useState<TagOptionItem[]>([]);
-  const [tagSelection, setTagSelection] = useState<TagSelection>({});
-  const totalActiveFilters = Object.keys(tagSelection).length;
+  const clauseInclude: string[] = useMemo(
```
Can we try using `id` instead of `name` for the `tagSelection`? It seems to me that it would be the cleanest approach and we wouldn't need all this logic here.

angorayc committed:
unit tests

angorayc committed:
Merge branch 'main' of github.com:elastic/kibana into dashboard-listing

angorayc committed:
revert changes for fixing active tag count

angorayc committed:
add tests

angorayc committed:
types

angorayc committed:
Merge branch 'main' into dashboard-listing

angorayc committed:
revert change

angorayc committed:
Merge branch 'main' of github.com:elastic/kibana into dashboard-listing

angorayc committed:
export dashboardListingTable

angorayc committed:
Merge branch 'main' of github.com:elastic/kibana into dashboard-listing

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

angorayc committed:
types

angorayc committed:
Merge branch 'dashboard-listing' of github.com:angorayc/kibana into dashboard-listing

angorayc committed:
test

sebelga commented on src/plugins/dashboard/public/dashboard_listing/dashboard_listing_table.tsx:
```diff
@@ -0,0 +1,102 @@
+            http,
+          },
+          toMountPoint,
+          savedObjectsTagging: savedObjectsTagging.hasApi // TODO: clean up this logic once https://github.com/elastic/kibana/issues/140433 is resolved
```
You need to put the changes I made to fix the tag list issue in Dashboard:
https://github.com/elastic/kibana/pull/160871/files#diff-aecc9bdefcbb1074958f2f7865f975e30c2b6cb24ca907c836df1596a04190ebR209

sebelga reviewed (changes requested):
Start to look good! 👍 One more change from my side to fix a big with the tag list. Cheers

semd commented on src/plugins/dashboard/public/dashboard_listing/hooks/use_dashboard_listing_table.tsx:
```diff
@@ -0,0 +1,272 @@
+  initialFilter,
+  urlStateEnabled,
+  useSessionStorageIntegration,
+}: {
+  dashboardListingId?: string;
+  disableCreateDashboardButton?: boolean;
+  getDashboardUrl: (dashboardId: string, usesTimeRestore: boolean) => string;
+  goToDashboard: (dashboardId?: string, viewMode?: ViewMode) => void;
+  headingId?: string;
+  initialFilter?: string;
+  urlStateEnabled?: boolean;
+  useSessionStorageIntegration?: boolean;
+}) => {
```
Do you think we can be explicit about the type the hook is expected to return here?

semd commented on src/plugins/dashboard/public/dashboard_listing/hooks/use_dashboard_listing_table.tsx:
```diff
@@ -0,0 +1,272 @@
+    [getDashboardUrl]
+  );
+
+  const tableListViewTableProps = useMemo(
```
We should set the type of this `tableListViewTableProps` object

semd commented on src/plugins/dashboard/public/dashboard_listing/dashboard_listing_table.tsx:
```diff
@@ -0,0 +1,102 @@
+  return (
+    <I18nProvider>
+      <TableListViewKibanaProvider
+        {...{
```
Can we memoize the `core` prop object?

ThomThomson reviewed (approved):
Presentation team changes LGTM! I tested locally and by looking through the code, and everything is working great! Nice work with splitting the Listing up into two components.

One thing that I noticed is that the Security Dashboard listing page has two `Create dashboard` buttons - is this expected?

Also left one small nit about the jest testing.

machadoum committed:
Implement review sugestions

angorayc committed:
Merge branch 'main' into dashboard-listing

angorayc committed:
review

angorayc committed:
review

angorayc commented on src/plugins/dashboard/public/dashboard_listing/hooks/use_dashboard_listing_table.test.tsx:
```diff
@@ -0,0 +1,274 @@
+  },
+}));
+
+const mockGetServices = {
```
We have an existing create dashboard button on the page and a new one in dashboard listing. I'll leave it as it is at the moment, as users might have used to the position of the existing button.

<img width="2545" alt="Screenshot 2023-07-18 at 10 08 50" src="https://github.com/elastic/kibana/assets/6295984/2041ba10-a885-40ba-b781-ae765da2e02c">


I've also updated the unit test and re-use the existing mocks. Thanks for the tips @ThomThomson !

sebelga reviewed (approved):
LGTM! 👍 Thanks for the patience and making those changes @angorayc !

machadoum commented on src/plugins/dashboard/public/dashboard_listing/dashboard_listing_table.tsx:
```diff
@@ -0,0 +1,112 @@
+
+  const savedObjectsTaggingFakePlugin = useMemo(
+    () =>
+      savedObjectsTagging.hasApi // TODO: clean up this logic once https://github.com/elastic/kibana/issues/140433 is resolved
```
Issue [140433](https://github.com/elastic/kibana/issues/140433) is closed. Is this comment outdated?

angorayc replied:
It's coming from https://github.com/elastic/kibana/pull/160540/files#diff-aecc9bdefcbb1074958f2f7865f975e30c2b6cb24ca907c836df1596a04190ebR62

@ThomThomson Should we remove the comment?

ThomThomson replied:
Nope, this comment is still valid - the linked issue is just the wrong way to solve it!

machadoum commented on src/plugins/dashboard/public/dashboard_listing/dashboard_listing_table.tsx:
```diff
@@ -0,0 +1,112 @@
+};
+
+// eslint-disable-next-line import/no-default-export
+export default DashboardListingTable;
```
Why the `export defaults` if it isn't lazy loaded?

angorayc replied:
It's a lazy loaded component: src/plugins/dashboard/public/dashboard_listing/index.tsx

machadoum commented on x-pack/plugins/security_solution/public/dashboards/pages/landing_page/index.tsx:
```diff
@@ -20,7 +20,25 @@ import * as i18n from './translations';
+  }
+  const uniqueQueryArray = securityTags?.reduce<string[]>((acc, { name }) => {
+    const nameString = `"${name}"`;
+    if (name && acc.indexOf(nameString) === -1) {
```
Very nit: Instead of checking if the value already exists in an array before pushing, you could use a [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set)

machadoum commented on x-pack/plugins/security_solution/public/dashboards/pages/landing_page/index.tsx:
```diff
@@ -20,7 +20,25 @@ import * as i18n from './translations';
+    return acc;
+  }, []);
+
+  const query = [uniqueQueryArray].join(' or');
```
I don't understand this logic. You are calling `join` on an array with only one element. Are the `[ ]` intended?

machadoum commented:
nit: I noticed it takes some seconds for the Dashboards table to load. Having a loading state/skeleton might be helpful for the user so they know that something is still loading.

https://github.com/elastic/kibana/assets/1490444/c6fdc7da-e707-4b2f-b485-23ab0b7d7042

I also noticed that the new tag never gets loaded when the [know issue](https://github.com/elastic/kibana/pull/160540#issuecomment-1610395834) happens and you navigate using the menu.

https://github.com/elastic/kibana/assets/1490444/ad215c14-0f5b-4cb7-a774-4bafbd1bd609

When I create a new dashboard it also doesn't show up:

https://github.com/elastic/kibana/assets/1490444/28f9952e-bf81-4907-a5c3-dba270d32cfe


The tag count is also wrong to me... 
<img width="407" alt="Screenshot 2023-07-20 at 12 35 39" src="https://github.com/elastic/kibana/assets/1490444/0622bd89-1825-48a0-b7cb-baedc90bfb52">


Am I doing something wrong? 🤔

angorayc committed:
Merge branch 'main' of github.com:elastic/kibana into dashboard-listing

angorayc committed:
use savedObjectsTagging client

angorayc committed:
unit test

angorayc committed:
review

angorayc commented:
@machadoum , thank you so much for testing this PR. I've added a loading icon and implemented the [recommended solution](https://github.com/elastic/kibana/issues/160723#issuecomment-1641904984) of https://github.com/elastic/kibana/issues/160723. Please have another try and see if it still happening, thanks again!

machadoum reviewed (approved):
I tried it out once more and it's working splendidly. Thank you, @angorayc!

angorayc commented:
https://github.com/elastic/kibana/pull/164476
